File: Services/pkg/role_permission/repository.go
package role_permission

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"

	"github.com/gofiber/fiber/v2"
	helpers "github.com/zercle/gofiber-helpers"
	"gorm.io/gorm"
)

type rolePermissionRepository struct {
	resources models.Resources
}

func NewRolePermissionRepository(resources models.Resources) domain.RolePermissionRepository {
	return &rolePermissionRepository{resources: resources}
}

func (r *rolePermissionRepository) Migrate() error {
	return r.resources.MainDbConn.AutoMigrate(&models.RolePermission{})
}

func (r *rolePermissionRepository) CreateRolePermission(ctx context.Context, rolePermission models.RolePermission) *helpers.ResponseError {
	if err := r.resources.MainDbConn.WithContext(ctx).Create(&rolePermission).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *rolePermissionRepository) GetRolePermission(ctx context.Context, id uint) (*models.RolePermission, *helpers.ResponseError) {
	var rolePermission models.RolePermission
	if err := r.resources.MainDbConn.WithContext(ctx).Where("id = ?", id).First(&rolePermission).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return &rolePermission, nil
}

func (r *rolePermissionRepository) GetRolePermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.RolePermission, *models.Pagination, *models.Search, *helpers.ResponseError) {
	var rolePermissions []models.RolePermission
	db := r.resources.MainDbConn.WithContext(ctx).Model(&models.RolePermission{})
	db = utils.ApplySearch(ctx, r.resources.FastHTTPClient, db, search, false)
	db = utils.ApplyPagination(db, &pagination, models.RolePermission{})
	if err := db.Find(&rolePermissions).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil, nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, nil, nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return rolePermissions, &pagination, &search, nil
}

func (r *rolePermissionRepository) GetRolePermissionsByRoleID(ctx context.Context, roleID uint) ([]models.RolePermission, *helpers.ResponseError) {
	var rolePermissions []models.RolePermission
	if err := r.resources.MainDbConn.WithContext(ctx).Where("role_id = ?", roleID).Find(&rolePermissions).Error; err != nil {
		return nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return rolePermissions, nil
}

func (r *rolePermissionRepository) UpdateRolePermission(ctx context.Context, id uint, rolePermission models.RolePermission) *helpers.ResponseError {
	if err := r.resources.MainDbConn.WithContext(ctx).Where("id = ?", id).First(&models.RolePermission{}).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := r.resources.MainDbConn.WithContext(ctx).Model(&models.RolePermission{}).Where("id = ?", id).Updates(rolePermission).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *rolePermissionRepository) DeleteRolePermission(ctx context.Context, id uint) *helpers.ResponseError {
	if err := r.resources.MainDbConn.WithContext(ctx).Where("id = ?", id).First(&models.RolePermission{}).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := r.resources.MainDbConn.WithContext(ctx).Delete(&models.RolePermission{}, id).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

func (r *rolePermissionRepository) DeleteRolePermissionsByRoleID(ctx context.Context, roleID uint) *helpers.ResponseError {
	if err := r.resources.MainDbConn.WithContext(ctx).Where("role_id = ?", roleID).Delete(&models.RolePermission{}).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return nil
}

File: hono-clean-architecture/pkg/role_permission/repository.ts
import { NodePgDatabase } from "drizzle-orm/node-postgres";
import { eq, and, sql, count } from "drizzle-orm";
import { rolePermissions, RolePermission, Pagination, Search, Resources } from "../models";
import { RolePermissionRepository as IRolePermissionRepository } from "../domain";
import { ResponseError, createError, whereAmI } from "../../internal/infrastructure/custom_error";
import { getOffset, getLimit } from "../utils";
import { RolePermissionCache, newRolePermissionCache } from "./cache";

export class RolePermissionRepository implements IRolePermissionRepository {
  private db: NodePgDatabase;
  private cache: RolePermissionCache;

  constructor(resources: Resources) {
    this.db = resources.mainDbConn;
    this.cache = newRolePermissionCache(resources.redisStorage);
  }

  async migrate(): Promise<void> {
    console.log("RolePermission table migration handled by Drizzle Kit");
  }

  async createRolePermission(data: {
    roleId: number;
    permissionId: number;
  }): Promise<ResponseError | null> {
    const existing = await this.db
      .select()
      .from(rolePermissions)
      .where(
        and(
          eq(rolePermissions.roleId, data.roleId),
          eq(rolePermissions.permissionId, data.permissionId)
        )
      )
      .limit(1);

    if (existing.length > 0) {
      return createError(409, whereAmI(), "Conflict", "RolePermission already exists.");
    }

    try {
      await this.db.insert(rolePermissions).values({
        roleId: data.roleId,
        permissionId: data.permissionId,
      });

      await this.cache.invalidateAllLists();
      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async getRolePermission(
    id: number
  ): Promise<{ rolePermission: RolePermission | null; error: ResponseError | null }> {
    const cached = await this.cache.get(id);
    if (cached) {
      return { rolePermission: cached, error: null };
    }

    try {
      const result = await this.db
        .select()
        .from(rolePermissions)
        .where(eq(rolePermissions.id, id))
        .limit(1);

      if (result.length === 0) {
        return {
          rolePermission: null,
          error: createError(404, whereAmI(), "Not Found", "RolePermission not found"),
        };
      }

      await this.cache.set(result[0]);
      return { rolePermission: result[0], error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        rolePermission: null,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async getRolePermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    rolePermissions: RolePermission[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }> {
    try {
      const countResult = await this.db.select({ count: count() }).from(rolePermissions);
      const total = Number(countResult[0]?.count ?? 0);

      let query = this.db.select().from(rolePermissions);

      if (pagination.orderBy) {
        const orderDirection = pagination.order === "desc" ? sql`DESC` : sql`ASC`;
        query = query.orderBy(
          sql`${sql.identifier(pagination.orderBy)} ${orderDirection}`
        ) as typeof query;
      }

      const result = await query.limit(getLimit(pagination)).offset(getOffset(pagination));

      return {
        rolePermissions: result,
        pagination: { ...pagination, total },
        search,
        error: null,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        rolePermissions: [],
        pagination,
        search,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async deleteRolePermission(id: number): Promise<ResponseError | null> {
    try {
      const rolePermission = await this.db
        .select()
        .from(rolePermissions)
        .where(eq(rolePermissions.id, id))
        .limit(1);

      if (rolePermission.length === 0) {
        return createError(404, whereAmI(), "Not Found", "RolePermission not found");
      }

      await this.db.delete(rolePermissions).where(eq(rolePermissions.id, id));
      await this.cache.invalidate(id);
      await this.cache.invalidateAllLists();

      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async getRolePermissionsByRoleId(
    roleId: number
  ): Promise<{ rolePermissions: RolePermission[]; error: ResponseError | null }> {
    try {
      const result = await this.db
        .select()
        .from(rolePermissions)
        .where(eq(rolePermissions.roleId, roleId));

      return { rolePermissions: result, error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        rolePermissions: [],
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }
}

export function newRolePermissionRepository(resources: Resources): IRolePermissionRepository {
  return new RolePermissionRepository(resources);
}
